platform/security/idp/security-idp-plugin/security-idp-plugin-audiencerestriction/src/test/groovy/org/codice/ddf/security/idp/plugin/audiencerestriction/AudienceRestrictionPresignPluginSpec.groovy

```diff
@@ -0,0 +1,131 @@
+    Conditions conditions = Mock(Conditions)
+    List<AudienceRestriction> audienceRestrictions = new ArrayList<>()
+    AudienceRestriction audienceRestriction = Mock(AudienceRestriction)
+    AudienceRestriction audienceRestriction2 = Mock(AudienceRestriction)
+    AudienceRestriction emptyAudienceRestriction = Mock(AudienceRestriction)
+    List<Audience> audiences = new ArrayList<>()
+    List<Audience> audiences2 = new ArrayList<>()
+    List<Audience> emptyAudiences = new ArrayList<>()
+    Audience audience = Mock(Audience)
+    Audience audience2 = Mock(Audience)
+
+    def setupSpec() {
+        OpenSAMLUtil.initSamlEngine()
+    }
+
+    def setup() {
+        // set up authNReq
+        authNReq.getIssuer() >> issuer
+        issuer.getValue() >> ISSUER_ID
+
+        // set up basic response
+        response.assertions >> assertions
+        assertions.add(assertion)
+        assertion.conditions >> conditions
+        conditions.audienceRestrictions >> audienceRestrictions
+
+        // set up audienceURIs
+        audience.audienceURI >> "https://localhost:8993/services/SecurityTokenService?wsdl"
+        audience2.audienceURI >> "https://localhost:8993/admin"
+
+        // add the audience to the audiences
+        audiences.add(audience)
+        audiences.add(audience2)
+        audiences2.add(audience)
+
+        // stub the audiences to the audienceRestriction
+        audienceRestriction.audiences >> audiences
+        audienceRestriction2.audiences >> audiences2
+        emptyAudienceRestriction.audiences >> emptyAudiences
+    }
+
+    def 'add audience to non existent audience restriction'() {
+        when:
+        plugin.processPresign(response, authNReq, spMetadata, bindings)
+
+        then:
+        response.assertions[0].conditions.audienceRestrictions[0].audiences.size() == 1
+        response.assertions[0].conditions.audienceRestrictions[0].audiences[0].audienceURI == ISSUER_ID
+    }
+
+    def 'add audience to audience restriction with no audiences'() {
+        setup:
+        // add empty audienceRestriction
+        audienceRestrictions.add(emptyAudienceRestriction)
+
+        when:
+        plugin.processPresign(response, authNReq, spMetadata, bindings)
+
+        then:
+        response.assertions[0].conditions.audienceRestrictions[0].audiences.size() == 1
+        response.assertions[0].conditions.audienceRestrictions[0].audiences[0].audienceURI == ISSUER_ID
+    }
+
+    def 'add audience to audience restriction with existing audiences'() {
+        setup:
+        // add audienceRestriction to audienceRestrictions
+        audienceRestrictions.add(audienceRestriction)
+
+        when:
+        plugin.processPresign(response, authNReq, spMetadata, bindings)
+
+        then:
+        response.assertions[0].conditions.audienceRestrictions[0].audiences.size() == 3
+        response.assertions[0].conditions.audienceRestrictions[0].audiences[2].audienceURI == ISSUER_ID
```

Comment:
✏️ This feels a little fragile, depending on the second entry in the audRest. list to be the value you expect. It's a bit cleaner to write this test like this:
```groovy
    def 'add audience to audience restriction with existing audiences'() {
        setup:
        // add audienceRestriction to audienceRestrictions
        audienceRestrictions.add(this.audienceRestriction)

        when:
        plugin.processPresign(response, authNReq, spMetadata, bindings)
        def audienceRestriction = response.assertions[0].conditions.audienceRestrictions[0]

        then:
        audienceRestriction.audiences.size() == 3
        audienceRestriction.audiences*.audienceURI.count {it == ISSUER_ID} == 1
    }
```
(Also note the `audienceRestriction` variable to simplify.)